drivers/input/touchscreen/himax.c

```diff
@@ -0,0 +1,462 @@
+			num_points_left--;
+	}
+
+	input_mt_sync_frame(ts->input_dev);
+	input_sync(ts->input_dev);
+}
+
+static bool himax_verify_checksum(struct himax_ts_data *ts,
+				  const struct himax_event *event)
+{
+	u8 *data = (u8 *)event;
+	int i;
+	u16 checksum = 0;
+
+	for (i = 0; i < sizeof(*event); i++)
+		checksum += data[i];
+
+	if ((checksum & 0x00ff) != 0) {
+		dev_err(&ts->client->dev, "Wrong event checksum: %04x\n",
+			checksum);
+		return false;
+	} else {
+		return true;
+	}
+}
+
+static void himax_handle_input(struct himax_ts_data *ts)
+{
+	int error;
+	struct himax_event event;
+
+	error = himax_read_input_event(ts, &event);
+	if (error) {
+		dev_err(&ts->client->dev, "Failed to read input event: %d\n",
+			error);
+		return;
+	}
+
+	if (!himax_verify_checksum(ts, &event))
+		return;
+
+	himax_process_event(ts, &event);
+}
+
+static irqreturn_t himax_irq_handler(int irq, void *dev_id)
+{
+	struct himax_ts_data *ts = dev_id;
+
+	himax_handle_input(ts);
+	return IRQ_HANDLED;
+}
+
+static int himax_request_irq(struct himax_ts_data *ts)
+{
+	struct i2c_client *client = ts->client;
+
+	return devm_request_threaded_irq(&client->dev, client->irq, NULL,
+					 himax_irq_handler,
+					 IRQF_TRIGGER_LOW | IRQF_ONESHOT,
+					 client->name, ts);
+}
+
+static int himax_probe(struct i2c_client *client,
+		       const struct i2c_device_id *id)
+{
+	int error;
+	struct device *dev = &client->dev;
+	struct himax_ts_data *ts;
+
+	if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C)) {
+		dev_err(dev, "%s: i2c check functionality error\n", __func__);
+		return -ENODEV;
+	}
+
+	ts = devm_kzalloc(dev, sizeof(*ts), GFP_KERNEL);
+	if (!ts)
+		return -ENOMEM;
+
+	i2c_set_clientdata(client, ts);
+	ts->client = client;
+
+	error = himax_setup_gpio(ts);
+	if (error)
+		return error;
+
+	error = himax_check_product_id(ts);
+	if (error)
+		return error;
+
+	himax_reset(ts);
+
+	error = himax_input_register(ts);
+	if (error) {
+		dev_err(dev, "%s: Unable to register input device\n", __func__);
+		return error;
+	}
+
+	error = himax_request_irq(ts);
+	if (error)
+		return error;
+
+	return 0;
+}
+
+static int himax_suspend(struct device *dev)
+{
+	struct himax_ts_data *ts = dev_get_drvdata(dev);
+
+	disable_irq(ts->client->irq);
+	return 0;
+}
+
+static int himax_resume(struct device *dev)
+{
+	struct himax_ts_data *ts = dev_get_drvdata(dev);
+
+	enable_irq(ts->client->irq);
+	return 0;
+}
+
+static SIMPLE_DEV_PM_OPS(himax_pm_ops, himax_suspend, himax_resume);
+
+// clang-format off
+static const struct i2c_device_id himax_ts_id[] = {
+	{ "hx83112a", 0 },
+	{ "hx83112b", 0 },
+	{ },
+};
+MODULE_DEVICE_TABLE(i2c, himax_ts_id);
+
+#ifdef CONFIG_OF
+static const struct of_device_id himax_of_match[] = {
+	{ .compatible = "himax,hx83112a" },
+	{ .compatible = "himax,hx83112b" },
+	{ },
+};
+MODULE_DEVICE_TABLE(of, himax_of_match);
+#endif
+// clang-format on
+
+static struct i2c_driver himax_ts_driver = {
+	.probe = himax_probe,
+	.id_table = himax_ts_id,
```

Comment:
you could even remove this id_table, it's not used in normal devicetree usecases, but it's also fine to keep it

Reply:
I guess that, in theory, the driver could be used without devicetree?

Reply:
Interestingly, the driver still gets loaded when I remove the entries from `himax_of_match` which means it uses `himax_ts_id` in that case I guess. It still gets its config from the devicetree though, I suppose. I don't see where else it could come from.

Reply:
Yes there's some weird fallback for dt also

Reply:
Would you agree with just keeping this for now?

Reply:
Ok